src/test/java/com/google/firebase/remoteconfig/FirebaseRemoteConfigClientImplTest.java

```diff
@@ -0,0 +1,377 @@
+                "Unexpected HTTP response with status: " + code + "\n{}");
+      }
+      checkGetRequestHeader(interceptor.getLastRequest());
+    }
+  }
+
+  @Test
+  public void testGetTemplateTransportError() {
+    client = initClientWithFaultyTransport();
+
+    try {
+      client.getTemplate();
+      fail("No error thrown for HTTP error");
+    } catch (FirebaseRemoteConfigException error) {
+      assertEquals(ErrorCode.UNKNOWN, error.getErrorCode());
+      assertEquals("Unknown error while making a remote service call: transport error",
+              error.getMessage());
+      assertTrue(error.getCause() instanceof IOException);
+      assertNull(error.getHttpResponse());
+      assertNull(error.getRemoteConfigErrorCode());
+    }
+  }
+
+  @Test
+  public void testGetTemplateSuccessResponseWithUnexpectedPayload() {
+    response.setContent("not valid json");
+
+    try {
+      client.getTemplate();
+      fail("No error thrown for malformed response");
+    } catch (FirebaseRemoteConfigException error) {
+      assertEquals(ErrorCode.UNKNOWN, error.getErrorCode());
+      assertTrue(error.getMessage().startsWith("Error while parsing HTTP response: "));
+      assertNotNull(error.getCause());
+      assertNotNull(error.getHttpResponse());
+      assertNull(error.getRemoteConfigErrorCode());
+    }
+    checkGetRequestHeader(interceptor.getLastRequest());
+  }
+
+  @Test
+  public void testGetTemplateErrorWithZeroContentResponse() {
+    for (int code : HTTP_ERRORS) {
+      response.setStatusCode(code).setZeroContent();
+
+      try {
+        client.getTemplate();
+        fail("No error thrown for HTTP error");
+      } catch (FirebaseRemoteConfigException error) {
+        checkExceptionFromHttpResponse(error, HTTP_2_ERROR.get(code), null,
+                "Unexpected HTTP response with status: " + code + "\nnull");
+      }
+      checkGetRequestHeader(interceptor.getLastRequest());
+    }
+  }
+
+  @Test
+  public void testGetTemplateErrorWithMalformedResponse() {
+    for (int code : HTTP_ERRORS) {
+      response.setStatusCode(code).setContent("not json");
+
+      try {
+        client.getTemplate();
+        fail("No error thrown for HTTP error");
+      } catch (FirebaseRemoteConfigException error) {
+        checkExceptionFromHttpResponse(error, HTTP_2_ERROR.get(code), null,
+                "Unexpected HTTP response with status: " + code + "\nnot json");
+      }
+      checkGetRequestHeader(interceptor.getLastRequest());
+    }
+  }
+
+  @Test
+  public void testGetTemplateErrorWithDetails() {
+    for (int code : HTTP_ERRORS) {
+      response.setStatusCode(code).setContent(
+              "{\"error\": {\"status\": \"INVALID_ARGUMENT\", \"message\": \"test error\"}}");
+
+      try {
+        client.getTemplate();
+        fail("No error thrown for HTTP error");
+      } catch (FirebaseRemoteConfigException error) {
+        checkExceptionFromHttpResponse(error, ErrorCode.INVALID_ARGUMENT);
```

Comment:
Also test for error message.

Reply:
As we call an overloaded `checkExceptionFromHttpResponse` version here, it checks for the default error message, `test error`.

Reply:
Actually, I will remove the overloaded method. Doesn't look like we need it anymore :)